packages/core/src/common/rpc.ts

```diff
@@ -0,0 +1,140 @@
+// *****************************************************************************
+// Copyright (C) 2022 Ericsson and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0.
+//
+// This Source Code may also be made available under the following Secondary
+// Licenses when the conditions for such availability set forth in the Eclipse
+// Public License v. 2.0 are satisfied: GNU General Public License, version 2
+// with the GNU Classpath Exception which is available at
+// https://www.gnu.org/software/classpath/license.html.
+//
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+// *****************************************************************************
+
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+import { inject, injectable } from 'inversify';
+import { CancellationToken } from './cancellation';
+import { Emitter, Event } from './event';
+import { Proxied, ProxyProvider } from './proxy';
+import { Reflection } from './reflection';
+import { serviceIdentifier } from './types';
+
+/**
+ * Represents a scoped connection to a remote service on which to call methods.
+ *
+ * There should be a 1-to-1 relationship between a `RpcConnection` and the
+ * remote service it represents.
+ */
+export const RpcConnection = serviceIdentifier<RpcConnection>('RpcConnection');
+export interface RpcConnection {
+    onClose: Event<void>
+    onRequest(handler: (method: string, params: any[], token: CancellationToken) => any): void
+    onNotification(handler: (method: string, params: any[]) => void): void
+    sendRequest<T>(method: string, params: any[]): Promise<T>
+    sendNotification(method: string, params: any[]): void
+}
```

Comment:
In my opinion this is one of the most important new abstractions in this change.

No matter how your RPC protocol works, or how you establish the connection, as long as you respect this interface we can create and serve JavaScript proxies.

Instances that implement this interface should be scoped to a service.

Reply:
IMO, we need to think about what we're abstracting here: is it transport (socket, electron, ipc) or is it lifecycle? I think the concepts we should deal with are service objects (proxies and servers) and transports (or channels in the related PR).  Everything in between should be an implementation detail.

Reply:
This is an implementation detail, but _remote_ procedure calls inherently carry the idea that there's some underlying "connection" to a remote. The connection related event `onClose` is only there to propagate to users of this API that the link is no longer valid and further calls won't work anymore. If the link never breaks, `onClose` will never be fired. Note that this API doesn't leak more than that to users, only notify proxies that RPC won't work anymore with the current instance.